core/extensions.rs

```diff
@@ -0,0 +1,141 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use crate::error::AnyError;
+use crate::{OpFn, OpId, OpState};
+
+pub type SourcePair = (&'static str, &'static str);
+pub type OpPair = (&'static str, Box<OpFn>);
+pub type OpMiddlewareFn = dyn Fn(&'static str, Box<OpFn>) -> Box<OpFn>;
+pub type OpStateFn = dyn Fn(&mut OpState) -> Result<(), AnyError>;
+
+#[derive(Default)]
+pub struct Extension {
+  js_files: Option<Vec<SourcePair>>,
+  ops: Option<Vec<OpPair>>,
+  opstate_fn: Option<Box<OpStateFn>>,
+  middleware_fn: Option<Box<OpMiddlewareFn>>,
+  initialized: bool,
+}
+
+impl Extension {
+  pub fn new(
+    js_files: Option<Vec<SourcePair>>,
+    ops: Option<Vec<OpPair>>,
+    opstate_fn: Option<Box<OpStateFn>>,
+    middleware_fn: Option<Box<OpMiddlewareFn>>,
+  ) -> Self {
+    Self {
+      js_files,
+      ops,
+      opstate_fn,
+      middleware_fn,
+      initialized: false,
+    }
+  }
+
+  pub fn pure_js(js_files: Vec<SourcePair>) -> Self {
+    Self::new(Some(js_files), None, None, None)
+  }
+
+  pub fn with_ops(
+    js_files: Vec<SourcePair>,
+    ops: Vec<OpPair>,
+    opstate_fn: Option<Box<OpStateFn>>,
+  ) -> Self {
+    Self::new(Some(js_files), Some(ops), opstate_fn, None)
+  }
+}
+
+// Note: this used to be a trait, but we "downgraded" it to a single concrete type
+// for the initial iteration, it will likely become a trait in the future
+impl Extension {
+  /// returns JS source code to be loaded into the isolate (either at snapshotting,
+  /// or at startup).  as a vector of a tuple of the file name, and the source code.
+  pub(crate) fn init_js(&self) -> Vec<SourcePair> {
+    match &self.js_files {
+      Some(files) => files.clone(),
+      None => vec![],
+    }
+  }
+
+  /// Called at JsRuntime startup to initialize ops in the isolate.
+  pub(crate) fn init_ops(&mut self, registrar: Rc<RefCell<dyn OpRegistrar>>) {
```

Comment:
Is there a reason this is `Rc<RefCell<dyn OpRegistrar>>`? Looking at the function I feel it could probably just be a plain `&mut dyn OpRegistrar` and do exactly the same. If this eventually becomes a trait then it feels a little dodgy giving it an `Rc<RefCell<_>>`since it would allow extensions to stash the `OpRegistrar` away and use it to dynamically create ops during code execution (at least, in theory) which feels like an odd thing to support? I might be wrong, it could be completely sensible to want that.

Reply:
I've killed the OpRegistrar trait and replaced it with first class op middleware.